File: Kotlin/src/main/kotlin/chapter4_factory/simple_factory/PizzaStore.kt
package chapter4_factory.simple_factory

class PizzaStore(
    private val factory: SimplePizzaFactory
) {
    fun orderPizza(type: String): Pizza {
        val pizza = factory.createPizza(type)
        pizza.prepare()
        pizza.bake()
        pizza.cut()
        pizza.box()
        return pizza
    }
}

File: Scala/src/main/scala/factory/PizzaStore.scala
package factory

import factory.PizzaType.PizzaType
import factory.ingredient.PizzaIngredientFactory
import factory.pizza.Pizza

trait PizzaStore {

  val ingredientFactory: PizzaIngredientFactory

  def orderPizza(pizzaType: PizzaType): Pizza = {
    val pizza = createPizza(pizzaType)
    pizza.prepare()
    pizza.bake()
    pizza.cut()
    pizza.box()
    pizza
  }

  def createPizza(pizzaType: PizzaType): Pizza

}
